### ISSUE TITLE
iFrame interaction issues

### ISSUE THREAD

**Issue by oiurch** [cypress-io/cypress#1763]
### Current behavior: 
I'm trying to interact with cross-origin iframe present on the website. 
In my test, when i don't have "chromeWebSecurity": false parameter set, all steps before the iframe interaction are passing and the interaction is not working (currently known issue).

But when i set the "chromeWebSecurity": false parameter in cypress.json my test start to fail on earlier steps like asserts, getting elements etc. and i can't even get to iframe interaction part.

As i've found in console, i've got error, like this:

Refused to display 'https://myurl.com/some_id' in a frame because it set 'X-Frame-Options' to 'sameorigin'

When the option isn't added in cypress.json i don't have such issues and all previous steps are working.
### Desired behavior:
I can interact with iFrame after adding "chromeWebSecurity": false (as suggested in discussions about iframes)
### Steps to reproduce:
Add "chromeWebSecurity": false to cypress.json and try to proceed to iframe and interact with it.
check the same without setting this parameter to false

```js

```

### Versions  
Cypress 2.1.0, Chrome66, Electron59

**Comment by pgroot91** [comment]
issue: #136 ?

**Comment by oiurch** [comment]
Yes, it's related, but adding "chromeWebSecurity": false isn't working for me, as it causes another issues, described above.

**Comment by Bkucera** [comment]
@oiurch set `chromeWebSecurity` in `cypress.json`, not `cypress.js`

**Comment by oiurch** [comment]
I've managed to pass steps previous to the iframe, but still looks like i can't interact with any of iframe elements with chromeWebSecurity set as false.
I'm trying it with this style:
```
cy.get('.view-content iframe').then(function ($iframe) {
      // This line works
      expect($iframe.attr('src')).to.contain('https://myurl.com');
      // Any attempt to get elements inside the iframe fails, e.g.:
      cy.get('span').contains('Categories').click();
    });
```

**Comment by itaykotler-fundbox** [comment]
@oiurch - I have found a way to bypass the iframe issues. Works with the latest Chrome verision 67.0.3396.62

In my `plugins/index.js` file I have added the following: 
```
module.exports = (on, config) => {
  on("before:browser:launch", (browser = {}, args) => {
    if (browser.name === "chrome") {
      args.push("--disable-features=CrossSiteDocumentBlockingIfIsolating,CrossSiteDocumentBlockingAlways,IsolateOrigins,site-per-process");
      args.push("--load-extension=cypress/extensions/Ignore-X-Frame-headers_v1.1");
      return args;
    }
  });
};
```
You can download the extension from [here](https://chrome-extension-downloader.com/).
Just copy-paste [this](https://chrome-extension-downloader.com/) and unpack it to your extension folder.
You can find more details about the disabled flags [here](https://www.chromium.org/Home/chromium-security/site-isolation).
Good luck

**Comment by 2RajShaikh** [comment]
Unfortunately I am still facing this issue. Is there any way to bypass this issue with any configuration flags?
**Refused to display** '_https://accounts.google.com/o/oauth2/v2/auth?client_id=....._' in a frame because it set '**X-Frame-Options**' to '**sameorigin**'.

**Comment by YOU54F** [comment]
> @oiurch - I have found a way to bypass the iframe issues. Works with the latest Chrome verision 67.0.3396.62
> 
> In my `plugins/index.js` file I have added the following:
> 
> ```js
> module.exports = (on, config) => {
>   on("before:browser:launch", (browser = {}, args) => {
>     if (browser.name === "chrome") {
>       args.push("--disable-features=CrossSiteDocumentBlockingIfIsolating,CrossSiteDocumentBlockingAlways,IsolateOrigins,site-per-process");
>       args.push("--load-extension=cypress/extensions/Ignore-X-Frame-headers_v1.1");
>       return args;
>     }
>   });
> };
> ```
> 
> You can download the extension from [here](https://chrome-extension-downloader.com/).
> Just copy-paste [this](https://chrome.google.com/webstore/detail/ignore-x-frame-headers/gleekbfjekiniecknbkamfmkohkpodhe) and unpack it to your extension folder.
> You can find more details about the disabled flags [here](https://www.chromium.org/Home/chromium-security/site-isolation).
> Good luck

This worked for me, you absolute legend @itaykotler-fundbox

**Comment by dhair-seva** [comment]
> @oiurch - I have found a way to bypass the iframe issues. Works with the latest Chrome verision 67.0.3396.62
> 
> In my `plugins/index.js` file I have added the following:
> 
> ```js
> module.exports = (on, config) => {
>   on("before:browser:launch", (browser = {}, args) => {
>     if (browser.name === "chrome") {
>       args.push("--disable-features=CrossSiteDocumentBlockingIfIsolating,CrossSiteDocumentBlockingAlways,IsolateOrigins,site-per-process");
>       args.push("--load-extension=cypress/extensions/Ignore-X-Frame-headers_v1.1");
>       return args;
>     }
>   });
> };
> ```
> 
> You can download the extension from [here](https://chrome-extension-downloader.com/).
> Just copy-paste [this](https://chrome.google.com/webstore/detail/ignore-x-frame-headers/gleekbfjekiniecknbkamfmkohkpodhe) and unpack it to your extension folder.
> You can find more details about the disabled flags [here](https://www.chromium.org/Home/chromium-security/site-isolation).
> Good luck

I cannot go to the link to the extension. Any other link that would work? I'm having the same issue

**Comment by dhair-seva** [comment]
I'm sad to report that this error came back after the computer restarted. This morning, I began to work on cypress tests again and now they are broken because this error came back, but I have not changed any of the cypress code (it still has the load extension code). Any ideas?

**Comment by itaykotler-fundbox** [comment]
@dhair-seva, make sure that "chromeWebSecurity" is set to `false` in `cypress.json`.
Maybe try updating Cypress and Chrome and let me know

**Comment by gowth90** [comment]
@itaykotler-fundbox 
Using the extension as plugin works in interactive mode. 
However, I am now looking for a way to handle the x-frame issue while running cypress in headless mode. Any ideas?

**Comment by itaykotler-fundbox** [comment]
@gowth90, this extension works well in our CI flow when running in headless. Can you add some more details?

**Comment by gowth90** [comment]
@itaykotler-fundbox , Thanks for the response. 
Headless runs in electron browser and in the below code works for chrome browser. Do I need to make any change to this code because headless runs in electron?
`module.exports = (on, config) => {
  on("before:browser:launch", (browser = {}, args) => {
   if (browser.name === "chrome") {
      args.push("--disable-features=CrossSiteDocumentBlockingIfIsolating,CrossSiteDocumentBlockingAlways,IsolateOrigins,site-per-process");
      args.push("--load-extension=cypress/extensions/Ignore-X-Frame-headers_v1.1");
      return args;
    }
  });
};`

I tried the below code, but no luck:
else if (browser.name === 'electron') {
       args['additionalArguments'] = ['--load-extension=cypress/extensions/Ignore-X-Frame-headers_v1.1', '--disable-features=CrossSiteDocumentBlockingIfIsolating,CrossSiteDocumentBlockingAlways,IsolateOrigins,site-per-process'];
}

So, basically I wanna know how to load extension with electron.

Thanks

**Comment by YOU54F** [comment]
You can’t load extensions in electron but you can run chrome in cypress run mode in ci

**Comment by itaykotler-fundbox** [comment]
@gowth90, as @YOU54F is correct.
You'll need to run cypress in CI mode using the `--browser` flag.
See [this](https://docs.cypress.io/guides/guides/command-line.html#cypress-run) for more details

**Comment by JasonFairchild** [comment]
Is it just implicit that one needs to pack an extension based of the manifest.json and background.js that you create inside a cypress/extensions folder? I think I'm just a total extension noob, so I failed to understand this until a colleague helped me out. But I thought I'd throw it out there in case any other like me comes along =). Once I did this, everything seemed to start working for me, but if I'm still doing something not quite right, I'd like to know. And thanks, this should really help me in my testing of salesforce. 

Maybe this is different if you get the extension directly from the chrome ext store, but I went the route of creating my own files and copying in exactly what @itaykotler-fundbox had posted for his files.

Basically do this, and then do 'Load unpacked' and choose the same directory as when you packed it:
![image](https://user-images.githubusercontent.com/56280950/70933883-f838a100-1ff9-11ea-9736-1dab81fdbd56.png)

**Comment by kesiyaabraham** [comment]
I tried the above-mentioned solution but no luck for me. I have extracted the CRX file and added it to the folder Ignore-X-Frame-headers_v1.1 and the code for plugins/index.js same as @itaykotler-fundbox. 

`module.exports = (on, config) => {
  on("before:browser:launch", (browser = {}, args) => {
    console.log(browser.name);
    if (browser.name === "chrome") {
      args.push("--disable-features=CrossSiteDocumentBlockingIfIsolating,CrossSiteDocumentBlockingAlways,IsolateOrigins,site-per-process");
      args.push("--load-extension=cypress/extensions/Ignore-X-Frame-headers_v1.1");
      return args;
    }
  });
};`

Also, set the chromewebSecurity to false in cypress.json
` "chromeWebSecurity":false,`

![image](https://user-images.githubusercontent.com/17702899/71445305-976a4080-273e-11ea-99b8-462804590654.png)

If I add the chrome extension 'Ignore X-Frame headers' directly to the browser from the web store where I run my tests, it is working fine.
Basically, I want to overcome the CORS exceptions in my testing site as it uses a different superdomain on the login page, that is why I set chromeWebSecurity to false. But after that, I was getting another issue 

"Refused to display 'XXXXXXXXX' in a frame because it set 'X-Frame-Options' to 'deny'.".

For this one, I added the chrome extension 'Ignore X frame headers' and it was working perfectly. But in the CI pipeline , I am getting the same issue and not able to add 'Ignore X frame headers' as I added it to my testing browser from web store.

Did I miss anything in the code? 

I am using the following versions,

Cypress:3.8.0
Chrome: Chrome 79

Thanks in advance for all the comments and answers!

**Comment by pratik-chakravorty** [comment]
I can't seem to run this in headless mode. I am using chrome as well.
My command is `cypress run --browser chrome --headless`
Headed mode works fine with no issues but as soon as I try headless it just fails with the same x-frame issue. Any ideas it would be really grand if someone can guide me here. Thanks

**Comment by Bkucera** [comment]
@pratik-chakravorty chrome does not support extensions in headless mode

**Comment by pratik-chakravorty** [comment]
@Bkucera So my best option is to run it in headed mode but will that work properly in CI builds. I am using Jenkins. I heard I need something called xvfb??

**Comment by JasonFairchild** [comment]
Is there a difference between chrome headless mode and Cypress CI mode? Sounds like others in Oct or before were running this extension in Cypress CI mode which I thought was referred to as headless. Or could this have to do with the difference before Cypress adopted the native chrome headless mode? Wasn't cypress using a kind of pseudo headless mode for chrome before then?

**Comment by jennifer-shehane** [comment]
Cypress runs on the specified browser given and each browser can be run headless or headed. Headless simply means that the browser is not visibly present when running the tests. I don't know what 'Cypress CI mode' is referring to, but by default `cypress run` runs in Electron headless. 

The headless mode with the Chrome browser was not available and working until 3.8.1.

You can always see the browser and mode that is being run in the output at the beginning of `cypress run` so there should be no confusion. 

<img width="334" alt="Screen Shot 2020-01-21 at 2 37 45 PM" src="https://user-images.githubusercontent.com/1271364/72786293-ab309600-3c5b-11ea-9bb9-ede9f9182758.png">

**Comment by manuelxaguilar** [comment]
@itaykotler-fundbox I followed your instructions on here but I am now getting all tests failing because of aborted XHR calls (namely Firestore calls). Any ideas?

@dhair-seva did you manage to solve your issues? Wondering if they were related maybe.

**Comment by JasonFairchild** [comment]
I got a nice error message in cypress about this anyway, but I though I would add a comment to hopefully help some avoid confusion. There is some syntax that is out of date in all the previous code examples. The changes should all be updated in this document: https://docs.cypress.io/api/plugins/browser-launch-api.html#Modify-browser-launch-arguments-preferences-and-extensions

But for this specific extension, this has been working for me (I don't think you have to provide the path to the extension if you pack and unpack it in the default folder for extensions for cypress?):
```
  on("before:browser:launch", (browser, launchOptions) => {
    if (browser.name === "chrome") {
      launchOptions.args.push(
        "--disable-features=CrossSiteDocumentBlockingIfIsolating,CrossSiteDocumentBlockingAlways,IsolateOrigins,site-per-process"
      );
      launchOptions.args.push(
        "--load-extension=cypress/extensions/Ignore-X-Frame-headers_v1.1"
      );
      return launchOptions;
    }
  });
```
